File: python/Unique_Binary_Search_Trees_II.py
"""
Given an integer n, generate all structurally unique BST's (binary search trees) that store values 1...n.

For example,
Given n = 3, your program should return all 5 unique BST's shown below.

   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3
"""
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def generateTrees(self, n):
        """
        :type n: int
        :rtype: List[TreeNode]
        """
        def generateTreesRange(start,end):
            res = []
            if start > end : res.append(None)
            elif start == end: res.append(TreeNode(start))
            else:
                for i in range(start, end+1):
                    left = generateTreesRange(start, i - 1)
                    right = generateTreesRange(i+1 , end)
                    for l in left:
                        for r in right:
                            root = TreeNode(i)
                            root.left = l
                            root.right = r
                            res.append(root)
            return res
        if n < 1 : return []
        return generateTreesRange(1,n)

File: c++/Unique_Binary_Search_Trees_II.cpp
/*
Given an integer n, generate all structurally unique BST's (binary search trees) that store values 1...n.

For example,
Given n = 3, your program should return all 5 unique BST's shown below.

   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3
   */
   /**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTreeRange(int start, int end){
        vector<TreeNode*> res;
        if(start > end) res.push_back(NULL);
        else if(start == end) res.push_back(new TreeNode(start));
        else{
            for(int i = start; i <= end ; i++){
                vector<TreeNode*> left = generateTreeRange(start, i-1);
                vector<TreeNode*> right = generateTreeRange(i+1, end);
                for(auto l : left){
                    for(auto r : right){
                        TreeNode* root = new TreeNode(i);
                        root->left = l;
                        root->right = r;
                        res.push_back(root);
                    }
                }
            }
        }
        return res;
    }
    vector<TreeNode*> generateTrees(int n) {
        vector<TreeNode*> res;
        if(n < 1) return res;
        return generateTreeRange(1,n);
    }
};
